### PR TITLE
Cross compilation + NPM publish

### PR DESCRIPTION
**Pull request by jackkleeman**
Adds npm packages for each target for cli and server, and base packages that import these using optional dependencies (https://blog.orhun.dev/packaging-rust-for-npm/)

Cross.toml had to be set with some particular env vars to get cross compiles working properly (and sccached), and also had to disabvle the experimental cmake build for rdkafka, and ensure libz is statically compiled there too

Will plumb this into the release process as a followup

### CODE REVIEW

File: Cross.toml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    "curl -OL https://github.com/protocolbuffers/protobuf/releases/download/v21.12/protoc-21.12-linux-x86_64.zip",
+    "unzip -o protoc-21.12-linux-x86_64.zip -d /usr/local bin/protoc",
+    "unzip -o protoc-21.12-linux-x86_64.zip -d /usr/local 'include/*'",
+    "rm -f protoc-21.12-linux-x86_64.zip",
+    "chmod +x /usr/local/bin/protoc",
+    "cd $(mktemp -d) && curl -LSfs https://github.com/mozilla/sccache/releases/download/v0.7.4/sccache-v0.7.4-x86_64-unknown-linux-musl.tar.gz -o sccache.tar.gz && tar -xvf sccache.tar.gz && rm sccache.tar.gz && cp sccache-v0.7.4-x86_64-unknown-linux-musl/sccache /usr/bin/sccache",
+]
+
+[target.aarch64-apple-darwin]
+image = "ghcr.io/jackkleeman/cross:aarch64-apple-darwin"
```
[the remark this answers is not in the document]
Comment (jackkleeman) [reply Cross.toml]: Built from here https://github.com/cross-rs/cross-toolchains/#targets

Contains osxcross https://github.com/tpoechtrager/osxcross to do macos cross compiles 
Similar to cockroachdb approach https://github.com/cockroachdb/cockroach/blob/master/build/toolchains/toolchainbuild/osxcross/perform-build.sh#L62

File: .github/workflows/binaries.yml [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            cp ../LICENSE "${node_pkg}"
+            # publish the package
+            pushd "${node_pkg}"
+            npm publish --access public
+            popd
+          done
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
+
+  publish-npm-base:
```
[the remark this answers is not in the document]
Comment (jackkleeman) [reply .github/workflows/binaries.yml:151]: base is `restate` which imports as optional dependencies `restate-darwin-arm64` and all other combinations

File: Cross.toml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    "curl -OL https://github.com/protocolbuffers/protobuf/releases/download/v21.12/protoc-21.12-linux-x86_64.zip",
+    "unzip -o protoc-21.12-linux-x86_64.zip -d /usr/local bin/protoc",
+    "unzip -o protoc-21.12-linux-x86_64.zip -d /usr/local 'include/*'",
+    "rm -f protoc-21.12-linux-x86_64.zip",
+    "chmod +x /usr/local/bin/protoc",
+    "cd $(mktemp -d) && curl -LSfs https://github.com/mozilla/sccache/releases/download/v0.7.4/sccache-v0.7.4-x86_64-unknown-linux-musl.tar.gz -o sccache.tar.gz && tar -xvf sccache.tar.gz && rm sccache.tar.gz && cp sccache-v0.7.4-x86_64-unknown-linux-musl/sccache /usr/bin/sccache",
+]
+
+[target.aarch64-apple-darwin]
+image = "ghcr.io/restatedev/cross:aarch64-apple-darwin"
```
[the remark this answers is not in the document]
Comment (jackkleeman) [reply Cross.toml:13]: cross does not publish arm images, aarch64 docker users should run with emulation (as i do). supporting an arm image would just add to our cross-compilation woes

File: justfile [hunk 4]
```diff
@@ -88,21 +88,8 @@ build-tools *flags: (_target-installed target)
...
-    elif [[ {{ target }} =~ "darwin" ]]; then
-      if [[ {{ os() }} != "macos" ]]; then
-        echo "Cannot built macos target on non-macos host";
-      else
-        just _resolved_target={{ target }} features={{ features }} build {{ flags }};
-      fi
-    else
-      echo "Unsupported target: {{ target }}";
-    fi
+    cross build --target {{ target }} {{ _features }} {{ flags }}
```
[the remark this answers is not in the document]
Comment (jackkleeman) [reply justfile:92]: does it not work for you? it works for me

[the remark this answers is not in the document]
Comment (tillrohrmann) [reply justfile:92]: It somewhat does but is so terribly slow that it is basically unusable.

[the remark this answers is not in the document]
Comment (jackkleeman) [reply justfile:92]: weird, its fine for me

[the remark this answers is not in the document]
Comment (tillrohrmann) [reply justfile:92]: Maybe it is my environment with Rancher Desktop?

[the remark this answers is not in the document]
Comment (jackkleeman) [reply justfile:92]: yeah, maybe - i am using orbstack which does pretty good virtualisation via rosetta

**Comment by jackkleeman** [comment]
@tillrohrmann should be good now. sadly doesn't reduce complexity much; we are still cross compiling to arm on x86 linux and mac builders, and indeed we still need to use `cross` on linux.

File: .github/workflows/binaries.yml [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+      - name: Generate notice
+        run: just notice-file
+
+      - name: Compile Mac binaries
+        if: ${{ matrix.build.os == 'macos-12' }}
+        uses: actions-rs/cargo@v1
+        env:
+          MACOSX_DEPLOYMENT_TARGET: 10.14.0
+          CC: sccache clang -target ${{ matrix.build.target }}
+          CXX: sccache clang++ -target ${{ matrix.build.target }}
```
Comment (tillrohrmann) [inline .github/workflows/binaries.yml:81]: Is this needed when building on macos? Differently asked, can the build step for linux and mac be exactly the same?

File: crates/ingress-kafka/Cargo.toml [hunk 6]
```diff
@@ -25,7 +25,7 @@ derive_builder = { workspace = true }
 drain = { workspace = true }
 opentelemetry_api = { workspace = true }
 prost = { workspace = true }
-rdkafka = { version = "0.34", features = ["cmake-build"] }
+rdkafka = { version = "0.34", features = ["libz-static"] }
```
Comment (tillrohrmann) [inline crates/ingress-kafka/Cargo.toml:28]: Is this needed for cross? If we are no longer using cross, should we revert this change?

**Review by tillrohrmann — COMMENTED** [review]
Thanks for creating this PR @jackkleeman. I've given it a pass and added some comments/questions.

File: .github/workflows/binaries.yml [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+      - uses: ./.github/actions/clean-runner
+
+      - name: Install Rust toolchain
+        uses: actions-rs/toolchain@v1
+        with:
+          # need to provide a toolchain version because this action cannot read rust-toolchain.toml. See https://github.com/actions-rs/toolchain/issues/126
+          toolchain: stable
+
+      # necessary because the toolchain@v1 action would install this target for stable only
```
[the remark this answers is not in the document]
Comment (jackkleeman) [reply .github/workflows/binaries.yml:49]: no, we are using 1.72

File: .github/workflows/binaries.yml [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        if: ${{ matrix.build.os == 'macos-12' }}
+        uses: actions-rs/cargo@v1
+        env:
+          MACOSX_DEPLOYMENT_TARGET: 10.14.0
+          CC: sccache clang -target ${{ matrix.build.target }}
+          CXX: sccache clang++ -target ${{ matrix.build.target }}
+        with:
+          command: build
+          args: --release --bins --target ${{ matrix.build.target }}
+          use-cross: ${{ matrix.build.os == 'ubuntu-22.04' }}
```
[the remark this answers is not in the document]
Comment (jackkleeman) [reply .github/workflows/binaries.yml:85]: this is just `false` - will fix

File: .github/workflows/binaries.yml [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    runs-on: ${{ matrix.build.os }}
+    strategy:
+      matrix:
+        build:
+          - target: aarch64-apple-darwin
+            os: macos-12
+          - target: x86_64-apple-darwin
+            os: macos-12
+          - target: x86_64-unknown-linux-gnu
+            os: ubuntu-22.04
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply .github/workflows/binaries.yml:31]: IMO Better to use stable targets, so we don't see the glibc changing from one build to another :)

File: .github/workflows/binaries.yml [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+      - name: Compile Linux binaries
+        if: ${{ matrix.build.os == 'ubuntu-22.04' }}
+        uses: actions-rs/cargo@v1
+        env:
+          RUSTC_WRAPPER: "" # Cross.toml sets it back
+        with:
+          command: build
+          args: --release --bins --target ${{ matrix.build.target }}
+          use-cross: true
```
[the remark this answers is not in the document]
Comment (jackkleeman) [reply .github/workflows/binaries.yml:95]: you need a cross compiler to build eg rocksdb, librdkafka for a different arch - clang IS a cross compiler, but gcc is not. i asked @slinkydeveloper and he wasnt able to build for arm with native cargo on his linux machine

File: justfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (jackkleeman) [reply justfile:92]: i no longer need this command so i will just revert

File: .github/workflows/binaries.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply .github/workflows/binaries.yml:95]: unfortunately that's the case. Every guide i stumbled on just proposed to use `cross`...

File: .github/workflows/binaries.yml [hunk 11]
```diff
@@ -0,0 +1,114 @@
+name: Build binaries
```
[the remark this answers is not in the document]
Comment (jackkleeman) [reply .github/workflows/binaries.yml:1]: will do this as a followup PR, but yes that would be the goal

File: .github/workflows/npm.yml [hunk 12]
```diff
@@ -0,0 +1,110 @@
+name: Publish npm binary packages
```
[the remark this answers is not in the document]
Comment (jackkleeman) [reply .github/workflows/npm.yml:1]: will handle in a followup; part of release process

File: .github/workflows/binaries.yml (hunk 5 above)
Comment (jackkleeman) [reply .github/workflows/binaries.yml:81]: this flag must be set for mac cross compile to work, otherwise librdkafka will get built for amd64. on linux runner, gcc is used

File: crates/ingress-kafka/Cargo.toml (hunk 6 above)
Comment (jackkleeman) [reply crates/ingress-kafka/Cargo.toml:28]: the cmake build for librdkafka is experimental and i found it highly unreliable on all targets when cross compiling. we still need cross on linux, and to cross compile on mac using clang also. ive synced with francesco and the cmake build isnt needed. and libz-static was being set implicitly in most builds anyway, just wanted to set it explicitly - we don't ever want to link against local libz

File: .github/workflows/binaries.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply .github/workflows/binaries.yml:1]: 👍

File: .github/workflows/binaries.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply .github/workflows/binaries.yml:49]: Thanks for the clarification. Maybe add a comment that the toolchain action before does not install stable but whatever is configured in the the toolchain.toml (1.72 atm).

File: .github/workflows/binaries.yml (hunk 5 above)
Comment (tillrohrmann) [reply .github/workflows/binaries.yml:81]: Maybe add this as a comment for future Restate generations.

File: .github/workflows/binaries.yml [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        if: ${{ matrix.build.os == 'macos-latest' }}
+        uses: actions-rs/cargo@v1
+        env:
+          MACOSX_DEPLOYMENT_TARGET: 10.14.0
+          CC: sccache clang -target ${{ matrix.build.target }}
+          CXX: sccache clang++ -target ${{ matrix.build.target }}
+        with:
+          command: build
+          args: --release --bins --target ${{ matrix.build.target }}
+          use-cross: false
```
Comment (tillrohrmann) [inline .github/workflows/binaries.yml:85]: Can we just use the default here (not specifying it)?

File: .github/workflows/binaries.yml (hunk 5 above)
Comment (tillrohrmann) [reply .github/workflows/binaries.yml:81]: Interesting. I am wondering why I can cross compile on my machine for x86 darwin because I haven't explicitly configured the target for clang and have not set up clang++ (at least I cannot remember).

File: .github/workflows/binaries.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply .github/workflows/binaries.yml:95]: Thanks for the clarification.

File: .github/workflows/npm.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply .github/workflows/npm.yml:1]: Great, thanks :-)

File: crates/ingress-kafka/Cargo.toml (hunk 6 above)
Comment (tillrohrmann) [reply crates/ingress-kafka/Cargo.toml:28]: Thanks for the clarification. Wondering why this has never been a problem on my machine also when cross compiling.

**Review by tillrohrmann — APPROVED** [review]
Thanks for updating this PR @jackkleeman. LGTM.

File: .github/workflows/binaries.yml (hunk 5 above)
Comment (jackkleeman) [reply .github/workflows/binaries.yml:81]: if we didnt want to use sccache we might be able to get away without these flags, but once I added the sccache prefix it didnt work without me setting -target. These things are very unpredictable...

File: crates/ingress-kafka/Cargo.toml (hunk 6 above)
Comment (jackkleeman) [reply crates/ingress-kafka/Cargo.toml:28]: the problem with the cmake build was that it looks around in your build environment to decide what to use and was making bad choices. this may be fine on your machine but i was finding it to work inconsistently in cross environments - we want the build process to be as predictable as possible, and here that seems to be the non cmake process

File: .github/workflows/binaries.yml (hunk 5 above)
Comment (jackkleeman) [reply .github/workflows/binaries.yml:81]: it may also be related to cmake/non-cmake builds; cmake made it vary a bit in outcome from environment to environment. without cmake, we have predictable outcomes, but we sometimes need to set flags
